### PR TITLE
[SPARK-5257] [MLlib] SparseVector indices must be non-negative

### PR DESCRIPTION
**Pull request by MechCoder**
Scala afaik does not support negative indexing. And fail early if negative indices are used.

### CODE REVIEW

**Comment by MechCoder** [comment]
@davies @mengxr This is a minor PR. Can you please review?

**Comment by davies** [comment]
@MechCoder Could you also do this for Python API?

**Comment by MechCoder** [comment]
Hmm.. If I do that, would it seem odd, if `SparseVector` in the PythonApi allows negative indexing, but not negative indices in the constructor?

I implemented negative indexing here, https://github.com/apache/spark/pull/4025 since Python and NumPy allow neg indexing, but that is not the case with Scala.

**Comment by MechCoder** [comment]
Thinking again, negative indexing makes sense, but not supplying an array of negative integers as indices. Let me know what you think about it.

**Comment by davies** [comment]
@MechCoder In #4025, you assume that all the indices are non-negative, so we need to add a check for that.

**Comment by MechCoder** [comment]
@davies Thanks. I've fixed it up. Anything more?

**Comment by MechCoder** [comment]
Sorry for the phony Jenkins comments. Is there a way I can turn it off, until I explicitly turn it on,

**Comment by MechCoder** [comment]
I am extremely sorry for the phony comments due to Jenkins. Is there any way to test it only when asked?

**Comment by srowen** [comment]
@MechCoder it tests every time the pull request is updated. that's fine and good since it always matters whether the most recent state passes tests. Just don't push changes to your branch until you're ready for Jenkins to test again.

File: python/pyspark/mllib/linalg.py [hunk 1]
```diff
@@ -335,6 +335,8 @@ def __init__(self, size, *args):
                 self.indices = np.array(args[0], dtype=np.int32)
                 self.values = np.array(args[1], dtype=np.float64)
             assert len(self.indices) == len(self.values), "index and value arrays not same length"
+            if np.any(self.indices < 0):
+                raise ValueError("indices should be positive.")
```
Comment (davies) [inline python/pyspark/mllib/linalg.py]: should it be non-negative?

**Comment by davies** [comment]
@MechCoder Looks good to me, just one minor comment.

**Comment by mengxr** [comment]
@MechCoder Similar to #3791, this will significantly hurt performance. Having indices being nonnegative and ordered is a contract. If you want to ensure these, please use the factory method `Vectors.sparse(size, entries)` to construct a sparse vector. Do you minding close this PR?

**Comment by MechCoder** [comment]
Alright, but maybe the documentation can be updated that the indices should be non-negative?

**Comment by MechCoder** [comment]
@mengxr The other PR that you refer to implements sorting, this however implements O(size) check for sparse vectors where size is typically really small (It does not support negative indices).

Even `scipy.sparse.coo_matrix` does a check here (https://github.com/scipy/scipy/blob/master/scipy/sparse/coo.py#L263) for negative row indices.

Please tell me if you still feel that this Pull Request is not that useful. I shall close it.

**Comment by mengxr** [comment]
@MechCoder This changes the construction time from O(1) to O(nnz), which could be thousands. We use many conversions from/to breeze. Adding this check would be too expensive. One option is to use the factory method with `entries`, or we can add another factory method: `sparse(size, indices, values, validate: Boolean)` if people feel unsafe without validation. But checking it every time would be too expensive.

**Comment by MechCoder** [comment]
Alright, closing please mark the issue as resolved.

**Pull request closed without merging**